---
  title: "IRE1a_poly_only_2023"
output: html_document
date: "2023-08-30"
---

```{r setup}
setwd("~/My Drive/Bench/IRE1 Polysome Seq 2022/")

library(data.table)
library(GenomicFeatures)
library(tximport)
library(DESeq2)
library(apeglm)
library(ggplot2)
library(ggrepel)
library(EnhancedVolcano)
library(rnaseqDTU)



```

###  METADATA
```{r metadata}
# Create metadata # WT_total_se_4 and IRE1aKO_total_se_3 are outliers and removed
metadata = data.frame(
  sample_id = c("IRE1aKO_poly_1", "IRE1aKO_poly_2", # IRE1aKO first polysome samples (batch 2)
                "IRE1aKO_poly_3.1", "IRE1aKO_poly_3.2", # IRE1aKO polysome heavy & light merged , sample 1, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_4.1", "IRE1aKO_poly_4.2", # IRE1aKO polysome heavy & light merged , sample 2 technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_5.1", "IRE1aKO_poly_5.2", # IRE1aKO polysome heavy & light merged , sample 3 technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_heavy_1.1", "IRE1aKO_poly_heavy_1.2", # IRE1aKO polysome heavy sample 1, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_heavy_2.1", "IRE1aKO_poly_heavy_2.2", # IRE1aKO polysome heavy sample 2, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_heavy_3.1", "IRE1aKO_poly_heavy_3.2", # IRE1aKO polysome heavy sample 3, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_light_1.1" ,"IRE1aKO_poly_light_1.2" , # IRE1aKO polysome light sample 1, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_light_2.1" ,"IRE1aKO_poly_light_2.2" , # IRE1aKO polysome light sample 2, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_poly_light_3.1", "IRE1aKO_poly_light_3.2", # IRE1aKO polysome light sample 3, technical replicates 1 & 2 (batch 3)
                "IRE1aKO_total_se_1","IRE1aKO_total_se_2","IRE1aKO_total_se_3", "IRE1aKO_total_se_4", # IRE1aKO original single end RNAseq (batch 1)
                "WT_poly_1", "WT_poly_2", # WT first polysome samples (batch 2)
                "WT_poly_3.1",  "WT_poly_3.2",  # WT polysome heavy & light merged , sample 1, technical replicates 1 & 2 (batch 3)
                "WT_poly_4.1",  "WT_poly_4.2",  # WT polysome heavy & light merged , sample 2, technical replicates 1 & 2 (batch 3)
                "WT_poly_heavy_1.1",  "WT_poly_heavy_1.2",  # WT polysome heavy sample 1, technical replicates 1 & 2 (batch 3)
                "WT_poly_heavy_2.1",  "WT_poly_heavy_2.2",  # WT polysome heavy sample 2, technical replicates 1 & 2 (batch 3)
                "WT_poly_light_1.1",  "WT_poly_light_1.2", # WT polysome light sample 1, technical replicates 1 & 2 (batch 3)
                "WT_poly_light_2.1",  "WT_poly_light_2.2", # WT polysome light sample 2, technical replicates 1 & 2 (batch 3)
                "WT_total_se_1", "WT_total_se_2","WT_total_se_3","WT_total_se_4" # WT original single end RNAseq (batch 1)
  ),
  sample = c("IRE1aKO_poly_1", "IRE1aKO_poly_2", # IRE1aKO first polysome samples (batch 2)
             "IRE1aKO_poly_3", "IRE1aKO_poly_3", # IRE1aKO polysome heavy & light merged , sample 1, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_4", "IRE1aKO_poly_4", # IRE1aKO polysome heavy & light merged , sample 2 technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_5", "IRE1aKO_poly_5", # IRE1aKO polysome heavy & light merged , sample 3 technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_heavy_1", "IRE1aKO_poly_heavy_1", # IRE1aKO polysome heavy sample 1, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_heavy_2", "IRE1aKO_poly_heavy_2", # IRE1aKO polysome heavy sample 2, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_heavy_3", "IRE1aKO_poly_heavy_3", # IRE1aKO polysome heavy sample 3, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_light_1" ,"IRE1aKO_poly_light_1" , # IRE1aKO polysome light sample 1, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_light_2" ,"IRE1aKO_poly_light_2" , # IRE1aKO polysome light sample 2, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_poly_light_3", "IRE1aKO_poly_light_3", # IRE1aKO polysome light sample 3, technical replicates 1 & 2 (batch 3)
             "IRE1aKO_total_se_1","IRE1aKO_total_se_2","IRE1aKO_total_se_3", "IRE1aKO_total_se_4", # IRE1aKO original single end RNAseq (batch 1)
             "WT_poly_1", "WT_poly_2", # WT first polysome samples (batch 2)
             "WT_poly_3",  "WT_poly_3",  # WT polysome heavy & light merged , sample 1, technical replicates 1 & 2 (batch 3)
             "WT_poly_4",  "WT_poly_4",  # WT polysome heavy & light merged , sample 2, technical replicates 1 & 2 (batch 3)
             "WT_poly_heavy_1",  "WT_poly_heavy_1",  # WT polysome heavy sample 1, technical replicates 1 & 2 (batch 3)
             "WT_poly_heavy_2",  "WT_poly_heavy_2",  # WT polysome heavy sample 2, technical replicates 1 & 2 (batch 3)
             "WT_poly_light_1",  "WT_poly_light_1", # WT polysome light sample 1, technical replicates 1 & 2 (batch 3)
             "WT_poly_light_2",  "WT_poly_light_2", # WT polysome light sample 2, technical replicates 1 & 2 (batch 3)
             "WT_total_se_1", "WT_total_se_2","WT_total_se_3" ,"WT_total_se_4" # WT original single end RNAseq (batch 1)
  ),
  condition = c(rep("IRE1aKO.polysome",8),
                rep("IRE1aKO.heavy", 6), 
                rep("IRE1aKO.light", 6),
                rep("IRE1aKO.total", 4),
                rep("WT.polysome", 6),
                rep("WT.heavy", 4), 
                rep("WT.light", 4),
                rep("WT.total", 4)
  ),
  genotype = c(rep("IRE1aKO",24), rep("WT",18)),
  fraction = c(rep("Polysome", 8), rep("Polysome.heavy",6),rep("Polysome.light", 6), rep("Total",4), 
               rep("Polysome", 6), rep("Polysome.heavy",4), rep("Polysome.light",4), rep("Total",4)),
  batch = c(rep("b2", 2), # IRE1aKO first polysome samples (batch 2)
            rep("b3", 6), # IRE1aKO polysome heavy & light merged (batch 3)
            rep("b3", 12), # IRE1aKO polysome heavy & light individual (batch 3)
            rep("b1", 4), # IRE1aKO original single end RNAseq
            rep("b2", 2), # WT first polysome samples (batch 2)
            rep("b3", 4), # WT polysome heavy & light merged (batch 3)
            rep("b3", 8), # WT polysome heavy & light individual (batch 3)
            rep("b1", 4) # WToriginal single end RNAseq
  ),
  technical_replicate = c("1","1","1","2","1","2","1","2", #  IRE1aKO Polysomes
                          "1","2","1","2","1","2", # IRE1aKO polysome heavy
                          "1","2","1","2","1","2", # IRE1aKO polysome light
                          "1","1","1","1",# IRE1aKO total se
                          "1","1","1","2","1","2", # WT polysomes
                          "1","2","1","2", # WT polysomes heavy
                          "1","2","1","2", # WT polysomes light
                          "1","1","1","1") # WT total se
)

# super important for how contrasts work
metadata$batch = factor(metadata$batch, levels = c("b1","b2","b3"))
metadata$genotype = factor(metadata$genotype, levels = c("WT","IRE1aKO"))
metadata$fraction = factor(metadata$fraction, levels = c("Total","Polysome","Polysome.heavy","Polysome.light"))
metadata$condition = factor(metadata$condition, 
                            levels = c("WT.total","WT.polysome","WT.heavy","WT.light",
                                       "IRE1aKO.total","IRE1aKO.polysome", "IRE1aKO.heavy","IRE1aKO.light"))


library(tximport)
library(readr)

### Transcript annotation
gtf <- "~/My Drive/Bench/IRE1 Polysome Seq 2022/annotations/gencode.vM33.annotation.gtf"  #GRCm39 vM33

txdb.filename <- "gencode.vM33.annotation.sqlite"

txdb <- makeTxDbFromGFF(gtf)

# Create a tx2gene dataframe 'txdf
library(GenomicFeatures)

txdf <- AnnotationDbi::select(txdb, keys(txdb, "TXNAME"), columns = "GENEID", keytype = "TXNAME")

### Salmon quant files
files <- file.path("salmon_quant", metadata$sample_id, "quant.sf")
names(files) <- list.files("salmon_quant")

txi <- tximport(files, type="salmon", txOut=TRUE, ignoreAfterBar = TRUE, # because salmon id has ENSMUST...|ENSMUSG...|genesymbol etc
                countsFromAbundance="scaledTPM", importer=read.delim) #,"dtuScaledTPM", tx2gene = txdf) 
cts <- txi$counts
cts <- cts[rowSums(cts) > 0,]

txdf <- txdf[match(rownames(cts),txdf$TXNAME),]
print("Can all TXNAME be matched?", quote = FALSE); all(rownames(cts) %in% txdf$TXNAME) # check if all can be mapped

counts <- data.frame(gene_id=txdf$GENEID,
                     feature_id=txdf$TXNAME,
                     cts)


######################################################################
########## Remove outliers from Counts & Metadata #########
drops <- c("IRE1aKO_total_se_3","WT_total_se_4")
counts <- counts[ , -which(names(counts) %in% drops) ]
metadata <- metadata[ !metadata$sample_id %in% drops , ]

######################################################################


counts$gene_id <- gsub(counts$gene_id, pattern="\\.[0-9]+$", replacement="") #remove suffixes so we can query biomart
counts$feature_id <- gsub(counts$feature_id, pattern="\\.[0-9]+$", replacement="") #remove suffixes so we can query biomart
# Lets annotate gene symbol so its included 
# Get mart
library(biomaRt)
ensembl <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
ATTRIBUTES <- c('ensembl_transcript_id','ensembl_gene_id', 'external_gene_name', 'gene_biotype', 'transcript_length')

# Fetch gene symbols
genes <- getBM(attributes = ATTRIBUTES, 
               filters = 'ensembl_transcript_id', 
               values = unique(counts$feature_id), 
               mart = ensembl)

# Merge with original count data
counts <- merge(x = counts, y = genes, 
                 by.x = "feature_id", 
                 by.y = "ensembl_transcript_id", 
                 all.x = TRUE)
library(dplyr)


counts$ensembl_gene_id <- counts$gene_id 
counts$gene_id <- counts$external_gene_name
counts <- counts %>% dplyr::select(gene_id, feature_id, ensembl_gene_id, external_gene_name, transcript_length, gene_biotype, everything())


###############################################################################################################

# Subset to protein coding genes only, (we want to look at splicing, it doesn't make sense to include lncRNA, we only lose power)
#counts <- subset(counts, gene_biotype =="protein_coding")

###############################################################################################################
#Manually collapse technical replicates ### should be functionalized later
# Counts are TPM, and technical replicates normally have minor variation and show poison, and should thus be averaged

counts$IRE1aKO_poly_3 <- (counts$IRE1aKO_poly_3.1 + counts$IRE1aKO_poly_3.2 ) /2
counts$IRE1aKO_poly_4 <- (counts$IRE1aKO_poly_4.1 + counts$IRE1aKO_poly_4.2 ) /2
counts$IRE1aKO_poly_5 <- (counts$IRE1aKO_poly_5.1 + counts$IRE1aKO_poly_5.2 ) /2

counts$IRE1aKO_poly_heavy_1 <- (counts$IRE1aKO_poly_heavy_1.1 + counts$IRE1aKO_poly_heavy_1.2 ) /2
counts$IRE1aKO_poly_heavy_2 <- (counts$IRE1aKO_poly_heavy_2.1 + counts$IRE1aKO_poly_heavy_2.2 ) /2
counts$IRE1aKO_poly_heavy_3 <- (counts$IRE1aKO_poly_heavy_3.1 + counts$IRE1aKO_poly_heavy_3.2 ) /2

counts$IRE1aKO_poly_light_1 <- (counts$IRE1aKO_poly_light_1.1 + counts$IRE1aKO_poly_light_1.2 ) /2
counts$IRE1aKO_poly_light_2 <- (counts$IRE1aKO_poly_light_2.1 + counts$IRE1aKO_poly_light_2.2 ) /2
counts$IRE1aKO_poly_light_3 <- (counts$IRE1aKO_poly_light_3.1 + counts$IRE1aKO_poly_light_3.2 ) /2

counts$WT_poly_3 <- (counts$WT_poly_3.1 + counts$WT_poly_3.2 ) /2
counts$WT_poly_4 <- (counts$WT_poly_4.1 + counts$WT_poly_4.2 ) /2

counts$WT_poly_heavy_1 <- (counts$WT_poly_heavy_1.1 + counts$WT_poly_heavy_1.2 ) /2
counts$WT_poly_heavy_2 <- (counts$WT_poly_heavy_2.1 + counts$WT_poly_heavy_2.2 ) /2

counts$WT_poly_light_1 <- (counts$WT_poly_light_1.1 + counts$WT_poly_light_1.2 ) /2
counts$WT_poly_light_2 <- (counts$WT_poly_light_2.1 + counts$WT_poly_light_2.2 ) /2

# Remove original columns ending in .1 or .2
counts <- counts[ , -grep("\\.1$|\\.2$", names(counts))]
# reorder names
counts <- counts[, order(names(counts))]
# move gene cols to beginning again
counts <- counts %>% dplyr::select(gene_id, feature_id, ensembl_gene_id, external_gene_name, transcript_length, everything())

###############################################################################################################

# make a collapsed metadata 
metadata.original <- metadata
metadata <- metadata[metadata$technical_replicate =="1",]
metadata$sample_id <- metadata$sample 

```



Total RNAseq was single end and  Polysome fraction was paired end in different batches so it doesn't make sense to analyse them together. We are now going to split the counts and metadata by experiment so that they can be analysed as separate objects. Statistically its also better to just use whole polysome rather than using light and heavy fractions which only have two replicates.

```{r split}
### metadata only poly and se
metadata.m  <- metadata[metadata$fraction %in% c("Total","Polysome"),]
rownames(metadata.m) <- metadata.m$sample_id
metadata.m$condition <- factor(metadata.m$condition, levels = c("WT.total","IRE1aKO.total","WT.polysome","IRE1aKO.polysome"))

####### Split off metadata & counts from total RNA seq experiment ###############
# Metadata, IRE1aKO_total_se_3 and WT_total_se_4 appear to be outliers, have much higher seq depth
metadata.se <- metadata[metadata$fraction == "Total",]
rownames(metadata.se) <- metadata.se$sample_id
#need to reset levels for condition, though we will only be using the genotype variable for testing
metadata.se$condition <- factor(metadata.se$condition, levels = c("WT.total","IRE1aKO.total"))

# Counts
counts.se <- counts[, c("gene_id","feature_id","ensembl_gene_id","external_gene_name","transcript_length","gene_biotype", rownames(metadata.se) ) ]




######### Split off metadata & counts from Polysome RNA seq experiments #########
# Metadata
metadata.poly <- metadata[metadata$fraction == "Polysome" ,]
rownames(metadata.poly) <- metadata.poly$sample_id
#need to reset levels for condition, though we will only be using the genotype variable for testing
metadata.poly$condition <- factor(metadata.poly$condition, levels = c("WT.polysome", "IRE1aKO.polysome"))

# Counts
counts.poly <- counts[, c("gene_id","feature_id","ensembl_gene_id","external_gene_name","transcript_length","gene_biotype", rownames(metadata.poly) ) ]


######### Split off metadata & counts from Polysome Heavy fraction RNA seq experiments
# Metadata
metadata.heavy <- metadata[metadata$fraction == "Polysome.heavy",]
rownames(metadata.heavy) <- metadata.heavy$sample_id
#need to reset levels for condition, though we will only be using the genotype variable for testing
metadata.heavy$condition <- factor(metadata.heavy$condition, levels = c("WT.heavy", "IRE1aKO.heavy"))
# Counts
counts.heavy <- counts[,c("gene_id","feature_id","ensembl_gene_id","external_gene_name","transcript_length","gene_biotype", rownames(metadata.heavy) ) ]


######### Split off metadata & counts from Polysome light fraction RNA seq experiments
# Metadata
metadata.light <- metadata[metadata$fraction == "Polysome.light",]
rownames(metadata.light) <- metadata.light$sample_id
#need to reset levels for condition, though we will only be using the genotype variable for testing
metadata.light$condition <- factor(metadata.light$condition, levels = c("WT.light", "IRE1aKO.light"))
# Counts
counts.light <- counts[,c("gene_id","feature_id","ensembl_gene_id","external_gene_name","transcript_length","gene_biotype", rownames(metadata.light) ) ]


plotMDS(counts[,7:length(counts)])
plotMDS(counts.se[,7:length(counts.se)])
plotMDS(counts.poly[,7:length(counts.poly)])
plotMDS(counts.heavy[,7:length(counts.heavy)])
plotMDS(counts.light[,7:length(counts.light)])


```



```{r prepare_drimseq_objects}

library(DRIMSeq)

counts.merged <- cbind(counts.se, counts.poly[,6:length(counts.poly)])

###### ALL SAMPLES TOGETHER, Polysomes merged # this would be for Genotype level test, so assume n.small to be at genotype level
d.all <- dmDSdata(counts=counts.merged, samples=metadata.m) # d is the imaginative name for our Drimseq oject
d.all

# Filter the data to remove lowly expressed transcripts 
# n should be number of samples, n.small, the smallest group
n <- length(counts.merged) - 6
n.small <- 7 
d.all <- dmFilter(d.all,
              min_samps_feature_expr=n.small, min_feature_expr=10,
              min_samps_feature_prop=n.small, min_feature_prop=0.1,
              min_samps_gene_expr=n, min_gene_expr=10)
d.all

# The dmDSdata object only contains genes that have more that one isoform, 
# which makes sense as we are testing for differential transcript usage. 
# We can find out how many of the remaining genes have N isoforms by tabulating 
# the number of times we see a gene ID, then tabulating the output again:
table(table(counts(d.all)$gene_id))





###### TOTAL RNASEQ DRIMSEQ OBJECT
d.se <- dmDSdata(counts=counts.se, samples=metadata.se) # d is the imaginative name for our Drimseq oject
d.se

# Filter the data to remove lowly expressed transcripts 
# n should be number of samples, n.small, the smallest group
n <- length(counts.se) - 5
n.small <- 3
d.se <- dmFilter(d.se,
              min_samps_feature_expr=n.small, min_feature_expr=10,
              min_samps_feature_prop=n.small, min_feature_prop=0.1,
              min_samps_gene_expr=n, min_gene_expr=10)
d.se

# The dmDSdata object only contains genes that have more that one isoform, 
# which makes sense as we are testing for differential transcript usage. 
# We can find out how many of the remaining genes have N isoforms by tabulating 
# the number of times we see a gene ID, then tabulating the output again:
table(table(counts(d.se)$gene_id))

###### POLYSOME DRIMSEQ OBJECT

d.poly <- dmDSdata(counts=counts.poly, samples=metadata.poly) # d is the imaginative name for our Drimseq oject
d.poly 

# Filter the data to remove lowly expressed transcripts 
# n should be number of samples, n.small, the smallest group
n <- length(counts.poly) - 5
n.small <- 4
d.poly <- dmFilter(d.poly,
              min_samps_feature_expr=n.small, min_feature_expr=10,
              min_samps_feature_prop=n.small, min_feature_prop=0.1,
              min_samps_gene_expr=n, min_gene_expr=10)
d.poly

# The dmDSdata object only contains genes that have more that one isoform, 
# which makes sense as we are testing for differential transcript usage. 
# We can find out how many of the remaining genes have N isoforms by tabulating 
# the number of times we see a gene ID, then tabulating the output again:
table(table(counts(d.poly)$gene_id))

###### POLYSOME HEAVY DRIMSEQ OBJECT
d.heavy <- dmDSdata(counts=counts.heavy, samples=metadata.heavy) # d is the imaginative name for our Drimseq oject
d.heavy

# Filter the data to remove lowly expressed transcripts 
# n should be number of samples, n.small, the smallest group
n <- length(counts.heavy) -5
n.small <- 2
d.heavy <- dmFilter(d.heavy,
              min_samps_feature_expr=n.small, min_feature_expr=10,
              min_samps_feature_prop=n.small, min_feature_prop=0.1,
              min_samps_gene_expr=n, min_gene_expr=10)
d.heavy

# The dmDSdata object only contains genes that have more that one isoform, 
# which makes sense as we are testing for differential transcript usage. 
# We can find out how many of the remaining genes have N isoforms by tabulating 
# the number of times we see a gene ID, then tabulating the output again:
table(table(counts(d.heavy)$gene_id))

###### POLYSOME LIGHT DRIMSEQ OBJECT
d.light <- dmDSdata(counts=counts.light, samples=metadata.light) # d is the imaginative name for our Drimseq oject
d.light

# Filter the data to remove lowly expressed transcripts 
# n should be number of samples, n.small, the smallest group
n <- length(counts.light) - 5
n.small <- 2
d.light <- dmFilter(d.light,
              min_samps_feature_expr=n.small, min_feature_expr=10,
              min_samps_feature_prop=n.small, min_feature_prop=0.1,
              min_samps_gene_expr=n, min_gene_expr=10)
d.light

# The dmDSdata object only contains genes that have more that one isoform, 
# which makes sense as we are testing for differential transcript usage. 
# We can find out how many of the remaining genes have N isoforms by tabulating 
# the number of times we see a gene ID, then tabulating the output again:
table(table(counts(d.light)$gene_id))


```



```{r DTUTesting_DRIMSEQ}
# Set multiple cores for parallel processing
BPPARAM <- BiocParallel::MulticoreParam(10)

# Create a model matrix, similar to how you would for DESeq2
# For total RNAseq
design_full.merged <- model.matrix(~batch + genotype, data=DRIMSeq::samples(d.all))
colnames(design_full.merged)
d.all <- dmPrecision(d.all, design=design_full.merged, BPPARAM = BPPARAM)
d.all <- dmFit(d.all, design=design_full.merged)
d.all <- dmTest(d.all, coef="genotypeIRE1aKO")


# Create a model matrix, similar to how you would for DESeq2
# For total RNAseq
design_full.se <- model.matrix(~genotype, data=DRIMSeq::samples(d.se))
colnames(design_full.se)
d.se <- dmPrecision(d.se, design=design_full.se, BPPARAM = BPPARAM)
d.se <- dmFit(d.se, design=design_full.se)
d.se <- dmTest(d.se, coef="genotypeIRE1aKO")

# For Polysome RNAseq
design_full.poly <- model.matrix(~batch + genotype, data=DRIMSeq::samples(d.poly))
colnames(design_full.poly)
d.poly <- dmPrecision(d.poly, design=design_full.poly, BPPARAM = BPPARAM)
d.poly <- dmFit(d.poly, design=design_full.poly)
d.poly <- dmTest(d.poly, coef="genotypeIRE1aKO")

# For Polysome heavy RNAseq
design_full.heavy <- model.matrix(~genotype, data=DRIMSeq::samples(d.heavy))
colnames(design_full.heavy)
d.heavy <- dmPrecision(d.heavy, design=design_full.heavy, BPPARAM = BPPARAM)
d.heavy <- dmFit(d.heavy, design=design_full.heavy)
d.heavy <- dmTest(d.heavy, coef="genotypeIRE1aKO")

# For Polysome light RNAseq
design_full.light <- model.matrix(~genotype, data=DRIMSeq::samples(d.light))
colnames(design_full.light)
d.light <- dmPrecision(d.light, design=design_full.light, BPPARAM = BPPARAM)
d.light <- dmFit(d.light, design=design_full.light)
d.light <- dmTest(d.light, coef="genotypeIRE1aKO")


# Extract results for all tests

all.res <- DRIMSeq::results(d.all)
head(all.res) 
all.res.txp <- DRIMSeq::results(d.all, level="feature")
head(all.res.txp)

se.res <- DRIMSeq::results(d.se)
head(se.res) 
se.res.txp <- DRIMSeq::results(d.se, level="feature")
head(se.res.txp)

poly.res <- DRIMSeq::results(d.poly)
head(poly.res)
poly.res.txp <- DRIMSeq::results(d.poly, level="feature")
head(poly.res.txp)

heavy.res <- DRIMSeq::results(d.heavy)
head(heavy.res)
heavy.res.txp <- DRIMSeq::results(d.heavy, level="feature")
head(heavy.res.txp)

light.res <- DRIMSeq::results(d.light)
head(light.res)
light.res.txp <- DRIMSeq::results(d.light, level="feature")
head(light.res.txp)

save.image(file = "IRE1aDRIMseqDTU.Rdata")

```


```{r plot_DRIMSEQ_DTU}

no.na <- function(x) ifelse(is.na(x), 1, x)

# Merged RNASeq
all.res$pvalue <- no.na(all.res$pvalue)
all.res.txp$pvalue <- no.na(all.res.txp$pvalue)

# Single end total RNASeq
se.res$pvalue <- no.na(se.res$pvalue)
se.res.txp$pvalue <- no.na(se.res.txp$pvalue)

# All Polysome
poly.res$pvalue <- no.na(poly.res$pvalue)
poly.res.txp$pvalue <- no.na(poly.res.txp$pvalue)

# Polysome Heavy Fraction
heavy.res$pvalue <- no.na(heavy.res$pvalue)
heavy.res.txp$pvalue <- no.na(heavy.res.txp$pvalue)

# Polysome Light Fraction
light.res$pvalue <- no.na(light.res$pvalue)
light.res.txp$pvalue <- no.na(light.res.txp$pvalue)


```
# Two stage filtering

From Love et al (2018): "A typical analysis of differential transcript usage would involve asking first: “which genes contain any evidence of DTU ?”, and secondly, “which transcripts in the genes that contain some evidence may be participating in the DTU?” Note that a gene may pass the first stage without exhibiting enough evidence to identify one or more transcripts that are participating in the DTU. The stageR package is designed to allow for such two-stage testing procedures, where the first stage is called a screening stage and the second stage a confirmation stage12. The methods are general, and can also be applied to testing, for example, changes across a time series followed by investigation of individual time points, as shown in the stageR package vignette. We show below how stageR is used to detect DTU and how to interpret its output."



```{r StageR}

# Function to prepare data for StageR:

ALPHA = 0.05
prep_and_StageR <- function(res, res.txp, prefix) {
  
  prepare_data <- function(res, res.txp, prefix) {
    # Ensure that 'res' has the required columns
    if (!all(c("pvalue", "gene_id") %in% names(res))) {
      stop("Input 'res' must have 'pvalue' and 'gene_id' columns.")
    }
    
    # Ensure that 'res.txp' has the required columns
    if (!all(c("pvalue", "feature_id", "gene_id") %in% names(res.txp))) {
      stop("Input 'res.txp' must have 'pvalue', 'feature_id', and 'gene_id' columns.")
    }
    
    # Extract and name pScreen from 'res'
    pScreen <- res$pvalue
    names(pScreen) <- res$gene_id
    
    # Create Confirmation matrix from 'res.txp'
    Confirmation <- matrix(res.txp$pvalue, ncol=1)
    rownames(Confirmation) <- gsub(res.txp$feature_id, pattern="\\.[0-9]+$", replacement="")
    
    # Arrange tx2gene data.frame
    tx2gene <- res.txp[,c("feature_id", "gene_id")]
    tx2gene$feature_id <- gsub(tx2gene$feature_id, pattern="\\.[0-9]+$", replacement="")
    
    # Return a list containing the three transformed objects, named with prefix
    output <- list(pScreen = pScreen, Confirmation = Confirmation, tx2gene = tx2gene)
    names(output) <- paste0(prefix, ".", names(output))
    
    return(output)
  }
  
  calculate_drim_padj <- function(pScreen, pConfirmation, prefix, tx2gene) {
    # Check if necessary libraries are installed
    if (!requireNamespace("stageR", quietly = TRUE)) {
      stop("You need to install the 'stageR' package")
    }
    
    # Load the library
    library(stageR)
    
    # Stage wise adjustment
    stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
                          pScreenAdjusted = FALSE, tx2gene = tx2gene)
    stageRObj <- stageWiseAdjustment(stageRObj, method = "dtu", alpha = ALPHA)
    
    # Get adjusted p values and suppress warnings
    drim_padj <- suppressWarnings({
      getAdjustedPValues(stageRObj, order = FALSE,
                         onlySignificantGenes = TRUE)
    })
    
    # Form output
    output_name <- paste(prefix, "drim.padj", sep = ".")
    assign(output_name, drim_padj, envir = .GlobalEnv)
    
    return(drim_padj)
  }
  
  # First function execution
  prepared_data <- prepare_data(res = res, res.txp = res.txp, prefix = prefix)
  
  # Extract the relevant results from prepared_data
  pScreen <- prepared_data[[paste0(prefix, ".pScreen")]]
  pConfirmation <- prepared_data[[paste0(prefix, ".Confirmation")]]
  tx2gene <- prepared_data[[paste0(prefix, ".tx2gene")]]
  
  # Second function execution
  drim_padj <- calculate_drim_padj(
    pScreen = pScreen, 
    pConfirmation = pConfirmation, 
    prefix = prefix, 
    tx2gene = tx2gene
  )
  
  # Optionally return everything in a list for further use
  results <- c(prepared_data, list(drim.padj = drim_padj))
  return(results)
}

# this function will output a prefix.drim.padj object
prep_and_StageR(res = all.res, res.txp=all.res.txp, prefix = "all")
prep_and_StageR(res = se.res, res.txp=se.res.txp, prefix = "se")
prep_and_StageR(res = poly.res, res.txp=poly.res.txp, prefix = "poly")
prep_and_StageR(res = heavy.res, res.txp=heavy.res.txp, prefix = "heavy")
prep_and_StageR(res = light.res, res.txp=light.res.txp, prefix = "light")

# Gene metadata
# Import information on transcript type 
transcript.type <- read.table("annotations/gencode.vM33.annotation_transcripts.txt", header = TRUE)
# Trim version numbers so we can merge with our tables
transcript.type$GeneID <- gsub(transcript.type$GeneID, pattern="\\.[0-9]+$", replacement="")
transcript.type$TranscriptID <- gsub(transcript.type$TranscriptID, pattern="\\.[0-9]+$", replacement="")
# Merge transcript length queried above by biomaRt
transcript.type <- merge(x= transcript.type , y= counts[,c("feature_id","transcript_length","gene_biotype")], by.x = "TranscriptID",by.y="feature_id")


se.res.txp <- merge(x = se.res.txp, y = transcript.type, by.x = "feature_id", by.y="TranscriptID")
all.res.txp <- merge(x = all.res.txp, y = transcript.type, by.x = "feature_id", by.y="TranscriptID")
poly.res.txp <- merge(x = poly.res.txp, y = transcript.type, by.x = "feature_id", by.y="TranscriptID")
heavy.res.txp <- merge(x = heavy.res.txp, y = transcript.type, by.x = "feature_id", by.y="TranscriptID")
light.res.txp <- merge(x = light.res.txp, y = transcript.type, by.x = "feature_id", by.y="TranscriptID")


poly.res <- poly.res[order(poly.res$pvalue, decreasing = FALSE), ]

plotProportions(d.poly, gene_id = "Aste1", "genotype", plot_type= "boxplot2")
plotProportions(d.poly, gene_id = "Flnb", "genotype", plot_type= "boxplot2")
plotProportions(d.poly, gene_id = "Rnaseh2a", "genotype", plot_type= "boxplot2")
plotProportions(d.heavy, gene_id = "Nhp2", "genotype", plot_type = "boxplot2")
plotProportions(d.heavy, gene_id = "Hnrnpf", "genotype", plot_type = "boxplot2")
plotProportions(d.poly, gene_id = "Hnrnpr", "genotype", plot_type = "boxplot2")
plotProportions(d.poly, gene_id = "R3hdm2", "genotype", plot_type = "boxplot2")
plotProportions(d.se, gene_id = "Chd9", "genotype", plot_type = "boxplot2")
plotProportions(d.se, gene_id = "Nhp2", "genotype", plot_type = "boxplot2")
plotProportions(d.se, gene_id = "Bcl11a", "genotype", plot_type = "boxplot2")
plotProportions(d.se, gene_id = "Ints1", "genotype", plot_type= "boxplot2")
plotProportions(d.heavy, gene_id = "Ints1", "genotype", plot_type= "boxplot2")
plotProportions(d.light, gene_id = "Ints1", "genotype", plot_type= "boxplot2")
plotProportions(d.se, gene_id = "Ints1", "genotype", plot_type= "boxplot2")
plotProportions(d.se, gene_id = "Pih1d2", "genotype", plot_type= "boxplot2")
plotProportions(d.poly, gene_id = "Pih1d2", "genotype", plot_type= "boxplot2")
plotProportions(d.heavy, gene_id = "Pih1d2", "genotype", plot_type= "boxplot2")
plotProportions(d.se, gene_id = "Eif4a1", "genotype", plot_type= "boxplot2")






save.image(file = "IRE1aDRIMseqDTU.Rdata")
```


# Differential Transcript Expression (DTE) using DESEQ2



```{r DTE}

# import with tximport 
# First subset files to remove outliers so it matches the filtered metadata

files.sub <- files[-which(names(files) %in% drops) ]

files.poly <- subset(files.sub, names(files.sub) %in% metadata.original[metadata.original$fraction %in% c("Polysome","Polysome.heavy","Polysome.light"),]$sample_id)

# from https://rdrr.io/github/bryancquach/omixjutsu/src/R/dge_utils.R
subset_txi <- function(txi, ids, dimension) {
  if (!dimension %in% c(1, 2)) {
    stop("Error: 'dimension' must be '1' (row) or '2' (column)")
  }
  index <- ids
  if (is.factor(index)) {
    index <- as.character(index)
  }
  if (is.character(index)) {
    if (dimension == 1) {
      if (sum(rownames(txi$counts) %in% index) != length(index)) {
        stop("IDs missing from txi object")
      }
      index <- match(x = index, table = rownames(txi$counts))
    } else {
      if (sum(colnames(txi$counts) %in% index) != length(index)) {
        stop("IDs missing from txi object")
      }
      index <- match(x = index, table = colnames(txi$counts))
    }
  }
  txi_subset <- txi
  if (dimension == 1) {
    txi_subset$counts <- txi_subset$counts[index, ]
    txi_subset$abundance <- txi_subset$abundance[index, ]
    txi_subset$length <- txi_subset$length[index, ]
  } else {
    txi_subset$counts <- txi_subset$counts[, index]
    txi_subset$abundance <- txi_subset$abundance[, index]
    txi_subset$length <- txi_subset$length[, index]
  }
  return(txi_subset)
}


txi.all.DTE <- tximport(files.sub, type="salmon", txIn = TRUE, txOut = TRUE, countsFromAbundance = "no", ignoreAfterBar=TRUE, importer=read.delim)
# trim suffixes so we can subset correctly
rownames(txi.all.DTE$counts) <- sub("\\..*", "", rownames(txi.all.DTE$counts))

txi.poly.DTE <- tximport(files.poly, type="salmon", txIn = TRUE, txOut = TRUE, countsFromAbundance = "no", ignoreAfterBar=TRUE, importer=read.delim)
# trim suffixes so we can subset correctly
rownames(txi.poly.DTE$counts) <- sub("\\..*", "", rownames(txi.poly.DTE$counts))

# Main dds object with batch included in design
dds.DTE <- DESeqDataSetFromTximport(txi.all.DTE, metadata.original, ~batch + genotype)
keep <- rowSums(counts(dds.DTE) >= 10) >= 3
dds.DTE <- dds.DTE[keep,]
dds.DTE <- collapseReplicates(dds.DTE, dds.DTE$sample, dds.DTE$technical_replicate )

# Main dds object from poly with batch and fraction included in design
dds.dte.poly.all <- DESeqDataSetFromTximport(txi.poly.DTE, metadata.original[metadata.original$fraction %in% c("Polysome","Polysome.heavy","Polysome.light"),], ~batch + fraction + genotype)
keep <- rowSums(counts(dds.dte.poly.all) >= 10) >= 3
dds.dte.poly.all <- dds.dte.poly.all[keep,]
dds.dte.poly.all <- collapseReplicates(dds.dte.poly.all, dds.dte.poly.all$sample, dds.dte.poly.all$technical_replicate )


# Main dds object with batch ommited from design 
dds.dte.nobatch <- DESeqDataSetFromTximport(txi.all.DTE, metadata.original, ~genotype)
keep <- rowSums(counts(dds.dte.nobatch) >= 10) >= 3
dds.dte.nobatch <- dds.dte.nobatch[keep,]
dds.dte.nobatch <- collapseReplicates(dds.dte.nobatch, dds.dte.nobatch$sample, dds.dte.nobatch$technical_replicate )

# Total RNAseq
dds.dte.se <- dds.dte.nobatch[,dds.dte.nobatch$fraction == "Total"]

# Polysome
dds.dte.poly <-  dds.DTE[,dds.DTE$fraction == "Polysome"]
dds.dte.poly$batch <- factor(dds.dte.poly$batch, levels = c("b2","b3")) #because batch 1 gets dropped need to fix levels

#Polysome Batch 3 only
dds.dte.poly.b3 <-  dds.dte.nobatch[,dds.dte.nobatch$batch == "b3" & dds.dte.nobatch$fraction == "Polysome"]
dds.dte.poly.b3$batch <- factor(dds.dte.poly.b3$batch, levels = c("b3")) #because batch 1 & 2 get dropped need to fix levels

#Polysome Heavy
dds.dte.heavy <-  dds.dte.nobatch[,dds.dte.nobatch$fraction == "Polysome.heavy"]

#Polysome Light
dds.dte.light <-  dds.dte.nobatch[,dds.dte.nobatch$fraction == "Polysome.light"]


dds.DTE <- DESeq(dds.DTE, parallel=TRUE)
dds.dte.se <- DESeq(dds.dte.se, parallel=TRUE,)
dds.dte.poly.all <- DESeq(dds.dte.poly.all, parallel=TRUE)
dds.dte.poly <- DESeq(dds.dte.poly, parallel=TRUE)
dds.dte.poly.b3 <- DESeq(dds.dte.poly.b3, parallel=TRUE)
dds.dte.heavy <- DESeq(dds.dte.heavy, parallel=TRUE)
dds.dte.light <- DESeq(dds.dte.light, parallel=TRUE)


res.dte.dds <- DESeq2::results(dds.DTE, name = "genotype_IRE1aKO_vs_WT")
res.dte.dds.LFC <- lfcShrink(dds.DTE, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds, ylim=c(-2,2)) 
DESeq2::plotMA(res.dte.dds.LFC, ylim=c(-2,2))

res.dte.dds.se <- DESeq2::results(dds.dte.se, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.se, ylim=c(-2,2)) 
res.dte.dds.se.LFC <- lfcShrink(dds.dte.se, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.LFC, ylim=c(-2,2))

res.dte.dds.poly.all <- DESeq2::results(dds.dte.poly.all, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.poly.all, ylim=c(-2,2)) 
res.dte.dds.poly.all.LFC <- lfcShrink(dds.dte.poly.all, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.poly.all.LFC, ylim=c(-2,2))

res.dte.dds.poly <- DESeq2::results(dds.dte.poly, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.poly, ylim=c(-2,2)) 
res.dte.dds.poly.LFC <- lfcShrink(dds.dte.poly, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.poly.LFC, ylim=c(-2,2))

res.dte.dds.poly.b3 <- DESeq2::results(dds.dte.poly.b3, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.poly.b3, ylim=c(-2,2))
res.dte.dds.poly.b3.LFC <- lfcShrink(dds.dte.poly, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.poly.b3.LFC, ylim=c(-2,2))

res.dte.dds.heavy <- DESeq2::results(dds.dte.heavy, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.heavy, ylim=c(-2,2))
res.dte.dds.heavy.LFC <- lfcShrink(dds.dte.heavy, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.heavy.LFC, ylim=c(-2,2))

res.dte.dds.light <- DESeq2::results(dds.dte.light, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dte.dds.light, ylim=c(-2,2))
res.dte.dds.light.LFC <- lfcShrink(dds.dte.light, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dte.dds.light.LFC, ylim=c(-2,2))

res.dte.dds.LFC <- as.data.frame(res.dte.dds.LFC)
res.dte.dds.LFC$TranscriptID <- rownames(res.dte.dds.LFC)
res.dte.dds.LFC <- merge(x = res.dte.dds.LFC, y = transcript.type, by ="TranscriptID")

res.dte.dds.se.LFC <- as.data.frame(res.dte.dds.se.LFC)
res.dte.dds.se.LFC$TranscriptID <- rownames(res.dte.dds.se.LFC)
res.dte.dds.se.LFC <- merge(x = res.dte.dds.se.LFC, y = transcript.type, by ="TranscriptID")

res.dte.dds.se <- as.data.frame(res.dte.dds.se)
res.dte.dds.se$TranscriptID <- rownames(res.dte.dds.se)
res.dte.dds.se <- merge(x = res.dte.dds.se, y = transcript.type, by ="TranscriptID")
write.table(res.dte.dds.se.LFC, file = "Total_RNA_DTE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dte.dds.poly.all.LFC <- as.data.frame(res.dte.dds.poly.all.LFC)
res.dte.dds.poly.all.LFC$TranscriptID <- rownames(res.dte.dds.poly.all.LFC)
res.dte.dds.poly.all.LFC <- merge(x = res.dte.dds.poly.all.LFC, y = transcript.type, by ="TranscriptID")
write.table(res.dte.dds.poly.all.LFC, file = "All_Polysome_DTE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dte.dds.poly.LFC <- as.data.frame(res.dte.dds.poly.LFC)
res.dte.dds.poly.LFC$TranscriptID <- rownames(res.dte.dds.poly.LFC)
res.dte.dds.poly.LFC <- merge(x = res.dte.dds.poly.LFC, y = transcript.type, by ="TranscriptID")
write.table(res.dte.dds.poly.LFC, file = "Polysome_DTE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dte.dds.poly.b3.LFC <- as.data.frame(res.dte.dds.poly.b3.LFC)
res.dte.dds.poly.b3.LFC$TranscriptID <- rownames(res.dte.dds.poly.b3.LFC)
res.dte.dds.poly.b3.LFC <- merge(x = res.dte.dds.poly.b3.LFC, y = transcript.type, by ="TranscriptID")

res.dte.dds.heavy.LFC <- as.data.frame(res.dte.dds.heavy.LFC)
res.dte.dds.heavy.LFC$TranscriptID <- rownames(res.dte.dds.heavy.LFC)
res.dte.dds.heavy.LFC <- merge(x = res.dte.dds.heavy.LFC, y = transcript.type, by ="TranscriptID")
write.table(res.dte.dds.heavy.LFC, file = "heavy_Polysome_DTE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dte.dds.light.LFC <- as.data.frame(res.dte.dds.light.LFC)
res.dte.dds.light.LFC$TranscriptID <- rownames(res.dte.dds.light.LFC)
res.dte.dds.light.LFC <- merge(x = res.dte.dds.light.LFC, y = transcript.type, by ="TranscriptID")
write.table(res.dte.dds.light.LFC, file = "light_Polysome_DTE.csv", quote = FALSE, sep = ",", row.names = FALSE)

library(ggplot2)

by_transcript_type <- ggplot(res.dte.dds.poly.all.LFC, aes(x= log2FoldChange, y = -log10(padj), color = Class) )+
  geom_point()+
  geom_text_repel(label = ifelse(abs(res.dte.dds.poly.all.LFC$log2FoldChange) > 4 & -log10(res.dte.dds.poly.all.LFC$padj) > 2.5, as.character(res.dte.dds.poly.all.LFC$TranscriptName), ''))
by_transcript_type

library(ggrepel)
box <- ggplot(res.dte.dds.poly.all.LFC, aes(x= Class, y = poly.log2FoldChange, color = Class) )+
  geom_boxplot()+
  geom_text_repel(label = ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > 3 & -log10(res.dte.dds.poly.all.LFC$poly.padj) > 2.5, as.character(res.dte.dds.poly.all.LFC$TranscriptName), ''))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position="none")
box

# lets merge different tables together for grouped plotting

colnames(res.dte.dds.se.LFC)[1:6] <- paste("se", colnames(res.dte.dds.se.LFC)[1:6], sep = ".")
colnames(res.dte.dds.poly.all.LFC)[1:6] <- paste("poly", colnames(res.dte.dds.poly.all.LFC)[1:6], sep = ".")
colnames(res.dte.dds.heavy.LFC)[1:6] <- paste("heavy", colnames(res.dte.dds.heavy.LFC)[1:6], sep = ".")
colnames(res.dte.dds.light.LFC)[1:6] <- paste("light", colnames(res.dte.dds.light.LFC)[1:6], sep = ".")


colnames(res.dte.dds.se.LFC)[colnames(res.dte.dds.se.LFC) == "se.TranscriptID"] <- "TranscriptID"
colnames(res.dte.dds.poly.all.LFC)[colnames(res.dte.dds.poly.all.LFC) == "poly.TranscriptID"] <- "TranscriptID"
colnames(res.dte.dds.heavy.LFC)[colnames(res.dte.dds.heavy.LFC) == "heavy.TranscriptID"] <- "TranscriptID"
colnames(res.dte.dds.light.LFC)[colnames(res.dte.dds.light.LFC) == "light.TranscriptID"] <- "TranscriptID"

# A series of left_joins
dte.df <- left_join(x = res.dte.dds.se.LFC[,1:6], 
                y = res.dte.dds.poly.all.LFC, 
                by = "TranscriptID")
dte.df <- left_join(x = dte.df, y = res.dte.dds.heavy.LFC[,1:6], by = "TranscriptID")
dte.df <- left_join(x = dte.df, y = res.dte.dds.light.LFC[,1:6], by = "TranscriptID")


se.box <- ggplot(dte.df, aes(x = Class, y = se.log2FoldChange, color = Class))+
  geom_boxplot()+
  geom_text(label = ifelse(abs(dte.df$se.log2FoldChange) > 6 &
                                   -log10(dte.df$se.padj)> 6,
                                 as.character(dte.df$GeneSymbol),''))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position="none")
se.box


se.violin <- ggplot(dte.df, aes(x = se.log2FoldChange, y = -log10(se.padj), color = Class))+
  geom_point()+
  geom_text(label = ifelse(abs(dte.df$se.log2FoldChange) > 6 &
                             dte.df$Class == "retained_intron" &
                                   -log10(dte.df$se.padj)> 6,
                                 as.character(dte.df$GeneSymbol),''))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position="none")
se.violin


################
poly.violin <- ggplot(dte.df, aes(x = poly.log2FoldChange, y = -log10(poly.padj), color = ifelse(abs(poly.log2FoldChange) > 1 &  poly.padj < 0.05, as.character(Class), NA)))+
  geom_point(size = 0.7, alpha= ifelse(abs(dte.df$poly.log2FoldChange) > 1 &  dte.df$poly.padj < 0.05, 1, 0.1))+
  geom_text_repel(label = ifelse(abs(dte.df$poly.log2FoldChange) > 1 &
                             dte.df$Class %in% c("nonsense_mediated_decay", "retained_intron", "snoRNA","snRNA") &
                                   dte.df$poly.padj < 0.05,
                                 as.character(dte.df$TranscriptName),''),
                  show.legend = FALSE)+
  # geom_text_repel(label = ifelse(abs(dte.df$poly.log2FoldChange) > 2 &
  #                            dte.df$Class %in% c("protein_coding") &
  #                                  dte.df$poly.padj < 0.01,
  #                                as.character(dte.df$GeneSymbol),''),
  #                 show.legend = FALSE)+
  labs(color=NULL)+
  scale_y_continuous(limits = c(0,10))+
  ggtitle(expression("Differential Transcript Expression in IRE1"*alpha*"KO Polysome RNA"))+
  xlab("log2 Fold Change")+
  ylab("-log10(Adjusted P value)")+
  coord_flip()+
  theme_minimal()
  
poly.violin
#####################


#### enhanced volcano of Poly DTE ####
# there are a couple very small p values with almost no log2FC that affect plotting, remove these

res.dte.dds.poly.all.LFC <- subset(res.dte.dds.poly.all.LFC, -log10(poly.padj) < 12.5)
rownames(res.dte.dds.poly.all.LFC) <- res.dte.dds.poly.all.LFC$TranscriptName

pCutoff = 0.05
FCcutoff = 0.8

lab_italics = paste0("italic('", rownames(res.dte.dds.poly.all.LFC), "')") 

selectlabels_italic <- paste0(
    "italic('",       
c("Hnrnpk-211", "Usp13-204", "Swi5-201", "Ints1-208", "Vps26a-201",  "Eif3d-205", "Eef2-203", "Snora61-201", "Snord13-201"), "')")


# paletteer_d("ggsci::springfield_simpsons")
# set colours for transcript types. apologies for the ridiculous syntax
 keyvals <- ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff &
    res.dte.dds.poly.all.LFC$Class == "protein_coding" , '#709AE1',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "retained_intron"   , '#FED439',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "nonsense_mediated_decay"   , '#F05C3B',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "lncRNA"   , '#D2AF81',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "processed_pseudogene"   , '#D5E4A2',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "processed_transcript"   , '#FD8CC1',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "protein_coding_CDS_not_defined"   , '#71D0F5',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "snoRNA"   , '#91331F',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "snRNA"   , '#46732E',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "unprocessed_pseudogene"   , '#370335',
      ifelse(abs(res.dte.dds.poly.all.LFC$poly.log2FoldChange) > FCcutoff &
                     res.dte.dds.poly.all.LFC$poly.padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "TEC"   , '#075149',
        'black')))))))))))
  keyvals[is.na(keyvals)] <- 'black'
  names(keyvals)[keyvals == '#709AE1'] <- 'protein coding'
  names(keyvals)[keyvals == '#FED439'] <- 'retained intron'
  names(keyvals)[keyvals == '#F05C3B'] <- 'nonsense mediated decay'
  names(keyvals)[keyvals == '#D2AF81'] <- 'lncRNA'
  names(keyvals)[keyvals == '#D5E4A2'] <- 'processed pseudogene'
  names(keyvals)[keyvals == '#FD8CC1'] <- 'processed transcript'
  names(keyvals)[keyvals == '#71D0F5'] <- 'protein coding CDS not defined'
  names(keyvals)[keyvals == '#91331F'] <- 'snoRNA'
  names(keyvals)[keyvals == '#46732E'] <- 'snRNA'
  names(keyvals)[keyvals == '#370335'] <- 'unprocessed pseudogene'
  names(keyvals)[keyvals == '#075149'] <- 'TEC'
  

dte.poly.volcano <- 
  EnhancedVolcano(res.dte.dds.poly.all.LFC, 
                  lab = lab_italics, 
                  selectLab = selectlabels_italic,
                  x = 'poly.log2FoldChange', y = 'poly.padj',
                  xlab = bquote(~Log[2]~ 'fold change (Shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  colCustom = keyvals,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff,
                  legendPosition = 'right',
                  pointSize = 2.0, 
                  labSize = 6.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Transcript Expression in IRE1"*alpha*"KO Total RNA"),
                  subtitle = "...",
                  caption = paste0('log2 FC cutoff: ', 
                    FCcutoff, '; p-value cutoff: ', 
                    pCutoff, '\nTotal = ', 
                    
                    nrow(res.dte.dds.poly.all.LFC), ' variables')
                  ) +coord_flip()#,
  #legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
  #legendPosition = 'right', legendLabSize = 14, legendIconSize = 5.0)
dte.poly.volcano


##################################################################



rownames(res.dte.dds.se) <- res.dte.dds.se$TranscriptName

pCutoff = 0.05
FCcutoff = 1

lab_italics = paste0("italic('", rownames(res.dte.dds.se), "')") 
selectlabels_italic <- paste0(
    "italic('",       
c("Hnrnpk-216", "Cux1-207", "Trmt2a-203", "Eif4enif1-201", "Eif4a2-213",  "Eif4a1-204", "Atrx-216 ", "Hes6-202", "Rps12-207","Nob1-204"), "')")


keyvals <- ifelse(abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff &
    res.dte.dds.poly.all.LFC$Class == "protein_coding" , '#709AE1',
      ifelse(abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "retained_intron"   , '#FED439',
      ifelse(abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "nonsense_mediated_decay"   , '#F05C3B',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "lncRNA"   , '#D2AF81',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "processed_pseudogene"   , '#D5E4A2',
      ifelse(abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "processed_transcript"   , '#FD8CC1',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "protein_coding_CDS_not_defined"   , '#71D0F5',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "snoRNA"   , '#91331F',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "snRNA"   , '#46732E',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "unprocessed_pseudogene"   , '#370335',
      ifelse( abs(res.dte.dds.se$log2FoldChange) > FCcutoff &
                     res.dte.dds.se$padj < pCutoff & res.dte.dds.poly.all.LFC$Class == "TEC"   , '#075149',
        'black')))))))))))
 
 
 
  keyvals[is.na(keyvals)] <- 'black'
  names(keyvals)[keyvals == '#709AE1'] <- 'protein coding'
  names(keyvals)[keyvals == '#FED439'] <- 'retained intron'
  names(keyvals)[keyvals == '#F05C3B'] <- 'nonsense mediated decay'
  names(keyvals)[keyvals == '#D2AF81'] <- 'lncRNA'
  names(keyvals)[keyvals == '#D5E4A2'] <- 'processed pseudogene'
  names(keyvals)[keyvals == '#FD8CC1'] <- 'processed transcript'
  names(keyvals)[keyvals == '#71D0F5'] <- 'protein coding CDS not defined'
  names(keyvals)[keyvals == '#91331F'] <- 'snoRNA'
  names(keyvals)[keyvals == '#46732E'] <- 'snRNA'
  names(keyvals)[keyvals == '#370335'] <- 'unprocessed pseudogene'
  names(keyvals)[keyvals == '#075149'] <- 'TEC'
  
dte.se.volcano <- 
  EnhancedVolcano(res.dte.dds.se, 
                  lab = lab_italics, 
                  selectLab = selectlabels_italic,
                  x = 'log2FoldChange', y = 'padj',
                  xlab = bquote(~Log[2]~ 'fold change (Shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  colCustom = keyvals,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff,
                  legendPosition = 'right',
                  pointSize = 2.0, 
                  labSize = 6.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Transcript Expression in IRE1"*alpha*"KO Polysome RNA"),
                  subtitle = "...",
                  caption = paste0('log2 FC cutoff: ', 
                    FCcutoff, '; p-value cutoff: ', 
                    pCutoff, '\nTotal = ', 
                    
                    nrow(res.dte.dds.se), ' variables')
                  ) +coord_flip()#,
  #legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
  #legendPosition = 'right', legendLabSize = 14, legendIconSize = 5.0)
dte.se.volcano


##################################################################



heavy.violin <- ggplot(dte.df, aes(x = heavy.log2FoldChange, y = -log10(heavy.padj), color = ifelse(abs(heavy.log2FoldChange) > 1 &  heavy.padj < 0.05, as.character(Class), NA)))+
  geom_point(size = 0.7, alpha= ifelse(abs(dte.df$heavy.log2FoldChange) > 1 &  dte.df$heavy.padj < 0.05, 1, 0.3))+
  geom_text_repel(label = ifelse(abs(dte.df$heavy.log2FoldChange) > 1 &
                             dte.df$Class %in% c("nonsense_mediated_decay", "retained_intron") &
                                   dte.df$heavy.padj < 0.05,
                                 as.character(dte.df$GeneSymbol),''))+
  labs(color=NULL)+
  theme_minimal()
  
heavy.violin

light.violin <- ggplot(dte.df, aes(x = light.log2FoldChange, y = -log10(light.padj), color = ifelse(abs(light.log2FoldChange) > 1 &  light.padj < 0.05, as.character(Class), NA)))+
  geom_point(size = 0.7, alpha= ifelse(abs(dte.df$light.log2FoldChange) > 1 &  dte.df$light.padj < 0.05, 1, 0.3))+
  geom_text_repel(label = ifelse(abs(dte.df$light.log2FoldChange) > 1 &
                             dte.df$Class %in% c("nonsense_mediated_decay", "retained_intron") &
                                   dte.df$light.padj < 0.05,
                                 as.character(dte.df$GeneSymbol),''))+
  labs(color=NULL)+
  theme_minimal()
  
light.violin



se.vs.poly.all <- ggplot(dte.df, aes(x = se.log2FoldChange, y = poly.log2FoldChange))+
  geom_point()+
  facet_wrap(~Class)
se.vs.poly.all 

poly.vs.heavy <- ggplot(dte.df, aes(x = poly.log2FoldChange, y = heavy.log2FoldChange))+
  geom_point(size = 0.5)+
  facet_wrap(~Class)+
  theme_minimal()+
  geom_text_repel(data= dte.df, label = ifelse(abs(dte.df$poly.log2FoldChange) > 2 & abs(dte.df$heavy.log2FoldChange) > 4, as.character(dte.df$GeneSymbol),''),
                  max.overlaps = 1000)
poly.vs.heavy


poly.vs.light <- ggplot(dte.df, aes(x = poly.log2FoldChange, y = light.log2FoldChange))+
  geom_point(size = 0.5)+
  facet_wrap(~Class)+
  theme_minimal()+
  geom_text_repel(data= dte.df, label = ifelse(abs(dte.df$poly.log2FoldChange) > 2 & abs(dte.df$light.log2FoldChange) > 4, as.character(dte.df$GeneSymbol),''),
                  max.overlaps = 1000)
poly.vs.light

retained.intron.df <- subset(dte.df, Class == "retained_intron")

se.vs.poly.intron <- ggplot(retained.intron.df, aes(x = poly.log2FoldChange, y = heavy.log2FoldChange))+
  geom_point()+
  geom_text_repel(label = ifelse(abs(retained.intron.df$poly.log2FoldChange) > 1 & abs(retained.intron.df$heavy.log2FoldChange) > 1, as.character(retained.intron.df$GeneSymbol),''))
se.vs.poly.intron



```

```{r plotDTE}






```







```{r DESEQ2}
# import with tximport and then convert transcripts to genes, which also calculates an average length per gene
# First subset files to remove outliers so it matches the filtered metadata, also subset to protein coding
files.sub <- files[-which(names(files) %in% drops) ]

files.poly <- subset(files.sub, names(files.sub) %in% metadata.original[metadata.original$fraction %in% c("Polysome","Polysome.heavy","Polysome.light"),]$sample_id)

# from https://rdrr.io/github/bryancquach/omixjutsu/src/R/dge_utils.R
subset_txi <- function(txi, ids, dimension) {
  if (!dimension %in% c(1, 2)) {
    stop("Error: 'dimension' must be '1' (row) or '2' (column)")
  }
  index <- ids
  if (is.factor(index)) {
    index <- as.character(index)
  }
  if (is.character(index)) {
    if (dimension == 1) {
      if (sum(rownames(txi$counts) %in% index) != length(index)) {
        stop("IDs missing from txi object")
      }
      index <- match(x = index, table = rownames(txi$counts))
    } else {
      if (sum(colnames(txi$counts) %in% index) != length(index)) {
        stop("IDs missing from txi object")
      }
      index <- match(x = index, table = colnames(txi$counts))
    }
  }
  txi_subset <- txi
  if (dimension == 1) {
    txi_subset$counts <- txi_subset$counts[index, ]
    txi_subset$abundance <- txi_subset$abundance[index, ]
    txi_subset$length <- txi_subset$length[index, ]
  } else {
    txi_subset$counts <- txi_subset$counts[, index]
    txi_subset$abundance <- txi_subset$abundance[, index]
    txi_subset$length <- txi_subset$length[, index]
  }
  return(txi_subset)
}


txi.all <- tximport(files.sub, type="salmon", tx2gene=txdf[,1:2], ignoreAfterBar=TRUE, importer=read.delim, countsFromAbundance = "lengthScaledTPM")
# trim suffixes so we can subset correctly
rownames(txi.all$counts) <- sub("\\..*", "", rownames(txi.all$counts))

txi.poly <- tximport(files.poly, type="salmon", tx2gene=txdf[,1:2], ignoreAfterBar=TRUE, importer=read.delim,
                     countsFromAbundance = "lengthScaledTPM")
# trim suffixes so we can subset correctly
rownames(txi.poly$counts) <- sub("\\..*", "", rownames(txi.poly$counts))

# Main dds object with batch included in design
dds <- DESeqDataSetFromTximport(txi.all, metadata.original, ~batch + genotype)
keep <- rowSums(counts(dds) >= 10) >= 3
dds <- dds[keep,]
dds <- collapseReplicates(dds, dds$sample, dds$technical_replicate )

# Main dds object from poly with batch and fraction included in design
dds.poly.all <- DESeqDataSetFromTximport(txi.poly, metadata.original[metadata.original$fraction %in% c("Polysome","Polysome.heavy","Polysome.light"),], ~batch + fraction + genotype)
keep <- rowSums(counts(dds.poly.all) >= 10) >= 3
dds.poly.all <- dds.poly.all[keep,]
dds.poly.all <- collapseReplicates(dds.poly.all, dds.poly.all$sample, dds.poly.all$technical_replicate )


# Main dds object with batch ommited from design 
dds.nobatch <- DESeqDataSetFromTximport(txi.all, metadata.original, ~genotype)
keep <- rowSums(counts(dds.nobatch) >= 10) >= 3
dds.nobatch <- dds.nobatch[keep,]
dds.nobatch <- collapseReplicates(dds.nobatch, dds.nobatch$sample, dds.nobatch$technical_replicate )

# Total RNAseq
dds.se <- dds.nobatch[,dds.nobatch$fraction == "Total"]

# Polysome
dds.poly <-  dds[,dds$fraction == "Polysome"]
dds.poly$batch <- factor(dds.poly$batch, levels = c("b2","b3")) #because batch 1 gets dropped need to fix levels

#Polysome Batch 3 only
dds.poly.b3 <-  dds.nobatch[,dds.nobatch$batch == "b3" & dds.nobatch$fraction == "Polysome"]
dds.poly.b3$batch <- factor(dds.poly.b3$batch, levels = c("b3")) #because batch 1 & 2 get dropped need to fix levels

#Polysome Heavy
dds.heavy <-  dds.nobatch[,dds.nobatch$fraction == "Polysome.heavy"]

#Polysome Light
dds.light <-  dds.nobatch[,dds.nobatch$fraction == "Polysome.light"]


dds <- DESeq(dds, parallel=TRUE)
dds.se <- DESeq(dds.se, parallel=TRUE,)
dds.poly.all <- DESeq(dds.poly.all, parallel=TRUE)
dds.poly <- DESeq(dds.poly, parallel=TRUE)
dds.poly.b3 <- DESeq(dds.poly.b3, parallel=TRUE)
dds.heavy <- DESeq(dds.heavy, parallel=TRUE)
dds.light <- DESeq(dds.light, parallel=TRUE)


res.dds <- DESeq2::results(dds, name = "genotype_IRE1aKO_vs_WT")
res.dds.LFC <- lfcShrink(dds, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds, ylim=c(-2,2)) 
DESeq2::plotMA(res.dds.LFC, ylim=c(-2,2))

res.dds.se <- DESeq2::results(dds.se, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.se, ylim=c(-2,2)) 
res.dds.se.LFC <- lfcShrink(dds.se, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.LFC, ylim=c(-2,2))

res.dds.poly.all <- DESeq2::results(dds.poly.all, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.poly.all, ylim=c(-2,2)) 
res.dds.poly.all.LFC <- lfcShrink(dds.poly.all, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.poly.all.LFC, ylim=c(-2,2))

res.dds.poly <- DESeq2::results(dds.poly, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.poly, ylim=c(-2,2)) 
res.dds.poly.LFC <- lfcShrink(dds.poly, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.poly.LFC, ylim=c(-2,2))

res.dds.poly.b3 <- DESeq2::results(dds.poly.b3, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.poly.b3, ylim=c(-2,2))
res.dds.poly.b3.LFC <- lfcShrink(dds.poly, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.poly.b3.LFC, ylim=c(-2,2))

res.dds.heavy <- DESeq2::results(dds.heavy, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.heavy, ylim=c(-2,2))
res.dds.heavy.LFC <- lfcShrink(dds.heavy, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.heavy.LFC, ylim=c(-2,2))

res.dds.light <- DESeq2::results(dds.light, name = "genotype_IRE1aKO_vs_WT")
DESeq2::plotMA(res.dds.light, ylim=c(-2,2))
res.dds.light.LFC <- lfcShrink(dds.light, coef="genotype_IRE1aKO_vs_WT", type="apeglm")
DESeq2::plotMA(res.dds.light.LFC, ylim=c(-2,2))



#annotate DESEQ2 results tables
library(biomaRt)

useEnsembl(biomart="genes",mirror="asia")
# Use biomaRt to get external gene names
getGeneNames <- function(ensembl_ids) {
  # Remove the suffixes
  ensembl_ids <- gsub("\\..*", "", ensembl_ids)
  mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", host="https://www.ensembl.org")
  genes <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name','gene_biotype'), 
                 filters = 'ensembl_gene_id', 
                 values = ensembl_ids, 
                 mart = mart)
  return(genes)
}


addGeneNamesToRes <- function(res) {
  gene_info <- getGeneNames(rownames(res))
  res$ensembl_gene_id <- gsub("\\..*", "", rownames(res)) # Remove suffixes for merging
  merged <- merge(res, gene_info, by = "ensembl_gene_id", all.x = TRUE)
  rownames(merged) <- merged$ensembl_gene_id
  merged$ensembl_gene_id <- NULL
  return(merged)
}



res.dds <- addGeneNamesToRes(as.data.frame(res.dds)) # these use biomaRt calls, so internet connection is required
res.dds.LFC <- addGeneNamesToRes(as.data.frame(res.dds.LFC))

res.dds.se <- addGeneNamesToRes(as.data.frame(res.dds.se))
res.dds.se.LFC <- addGeneNamesToRes(as.data.frame(res.dds.se.LFC))
res.dds.se.LFC$ensembl_gene_id <- rownames(res.dds.se.LFC)
write.table(res.dds.se.LFC, file = "Total_RNA_DGE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dds.poly.all <- addGeneNamesToRes(as.data.frame(res.dds.poly.all))
res.dds.poly.all.LFC <- addGeneNamesToRes(as.data.frame(res.dds.poly.all.LFC))
res.dds.poly.all.LFC$ensembl_gene_id <- rownames(res.dds.poly.all.LFC)
write.table(res.dds.poly.all.LFC, file = "All_Polysome_fractions_DGE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dds.poly <- addGeneNamesToRes(as.data.frame(res.dds.poly))
res.dds.poly.LFC <- addGeneNamesToRes(as.data.frame(res.dds.poly.LFC))
res.dds.poly.LFC$ensembl_gene_id <- rownames(res.dds.poly.LFC)
write.table(res.dds.poly.LFC, file = "Polysome_DGE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dds.poly.b3 <- addGeneNamesToRes(as.data.frame(res.dds.poly.b3))
res.dds.poly.b3.LFC <- addGeneNamesToRes(as.data.frame(res.dds.poly.b3.LFC))

res.dds.heavy <- addGeneNamesToRes(as.data.frame(res.dds.heavy))
res.dds.heavy.LFC <- addGeneNamesToRes(as.data.frame(res.dds.heavy.LFC))
res.dds.heavy.LFC$ensembl_gene_id <- rownames(res.dds.heavy.LFC)
write.table(res.dds.heavy.LFC, file = "Heavy_Polysome_DGE.csv", quote = FALSE, sep = ",", row.names = FALSE)

res.dds.light <- addGeneNamesToRes(as.data.frame(res.dds.light))
res.dds.light.LFC <- addGeneNamesToRes(as.data.frame(res.dds.light.LFC))
res.dds.light.LFC$ensembl_gene_id <- rownames(res.dds.light.LFC)
write.table(res.dds.light.LFC, file = "Light_Polysome_DGE.csv", quote = FALSE, sep = ",", row.names = FALSE)


# extract significant genes from each test

extractSignificantGenes <- function(res, padj_threshold = 0.05, lfcthreshold = 0.5) {
  significant_genes <- subset(res, padj < padj_threshold & 
                                abs(log2FoldChange) > lfcthreshold 
  ) 
  return(data.frame(ensembl_id = rownames(significant_genes), 
                    external_gene_name = significant_genes$external_gene_name))
}

extractSignificantGenesPC <- function(res, padj_threshold = 0.1, lfcthreshold = 0.3) {
  significant_genes <- subset(res, padj < padj_threshold & 
                                abs(log2FoldChange) > lfcthreshold &
                                gene_biotype == "protein_coding")
  return(data.frame(ensembl_id = rownames(significant_genes), 
                    external_gene_name = significant_genes$external_gene_name))
}



# This will now generate a list of data frames, with both IDs
list_significant_genes <- list( 
  all = extractSignificantGenes(res.dds),
  se = extractSignificantGenes(res.dds.se),
  poly.all = extractSignificantGenes(res.dds.poly.all),
  poly = extractSignificantGenes(res.dds.poly),
  poly.b3 = extractSignificantGenes(res.dds.poly.b3),
  heavy = extractSignificantGenes(res.dds.heavy),
  light = extractSignificantGenes(res.dds.light)
)


sig.list <- bind_rows(list_significant_genes) %>% distinct()

sig.poly <- bind_rows(list_significant_genes$poly.all, list_significant_genes$heavy, list_significant_genes$light, list_significant_genes$poly.b3) %>% distinct()


```





```{r heatmap}

library(DESeq2)
library(limma)
library(ComplexHeatmap)


######## Lets do it without total RNAseq, with collapsed technical replicates, where batch and fraction are included in design and in removeBatchEffect for matrix preparation collapsed technical replicates ###############
dds.heat.b <-  dds.poly.all

#Because we stripped ensembl_id suffixes above, we have to do the same in the dds object to match them
original_rownames.b <- rownames(dds.heat.b) #save them in case needed
rownames(dds.heat.b) <- gsub("\\..*", "", rownames(dds.heat.b))

# Subset VST-transformed data to only include the significant genes
vst_data.b <- assay(vst(dds.heat.b))
vst_data.b <- vst_data.b[sig.poly$ensembl_id, ]

# Replace the rownames of the matrix with external gene name
rownames(vst_data.b) <- sig.poly$external_gene_name

#create subset of metadata without total RNASeq
metadata_ordered.b <- metadata[metadata$batch %in% c("b2","b3") & metadata$technical_replicate =="1",]

metadata_ordered.b$batch <- factor(metadata_ordered.b$batch, levels = c("b2","b3")) 
metadata_ordered.b$fraction <- factor(metadata_ordered.b$fraction, levels = c("Polysome","Polysome.heavy","Polysome.light")) 
metadata_ordered.b$sample_id <- metadata_ordered.b$sample

# Remove fraction batch effect
vst_nobatch.b <- removeBatchEffect(vst_data.b, batch=metadata_ordered.b$batch)

# Scale the data
mat.b <- t(scale(t(vst_nobatch.b), scale = T))

# Ensure the columns in metadata correspond to the columns in the expression data
col_order.b <- colnames(mat.b)
metadata_ordered.b <- metadata_ordered.b[match(col_order.b, metadata_ordered.b$sample_id), ]
metadata_ordered.b <- subset(metadata_ordered.b, select= -c(sample_id, sample, technical_replicate, condition))




# Prepare annotations for the heatmap
ha.b = HeatmapAnnotation(df = metadata_ordered.b,
                         col = list(genotype = c("WT" = "tan","IRE1aKO"="#3690AE"),
                                    fraction = c("Polysome"= "#D04544",
                                                 "Polysome.heavy"="#33608C",
                                                 "Polysome.light"="#F5B35E"),
                                    batch = c("b2"= "#69737B","b3"="#C1C4C3")
                                    ),
                         gp = gpar(fontsize = 12,  fontface = 'bold')
                         )

require(circlize)
require(RColorBrewer)
require(cluster)
pamClusters <- cluster::pam(mat.b, k = 2) # pre-select k = 2 clusters
pamClusters$clustering <- paste0('Cluster ', pamClusters$clustering)

# fix order of the clusters to have 1 to 2, top to bottom
pamClusters$clustering <- factor(pamClusters$clustering, levels = c('Cluster 1', 'Cluster 2'))


myCol <- colorRampPalette(c('dodgerblue', 'black', 'yellow'))(100)
myBreaks <- seq(-3, 3, length.out = 100)

# 'Step through' gene annotations to only label every 10th row
genelabels <- rowAnnotation(
  Genes = anno_mark(
    at = seq(1, nrow(mat.b), 5),
    labels = rownames(mat.b)[seq(1, nrow(mat.b), 5)],
    labels_gp = gpar(fontsize = 12, fontface = 'italic'),
    padding = 0.5),
  width = unit(2.0, 'cm') +
    
    max_text_width(
      rownames(mat.b)[seq(1, nrow(mat.b), 5)],
      gp = gpar(fontsize = 12,  fontface = 'italic')))


# Plot the heatmap
heatmap.b = Heatmap(
  mat.b,
  #split the genes / rows according to the PAM clusters
  #split = pamClusters$clustering,
  #cluster_row_slices = TRUE,
  name = 'Gene\nZ-\nscore', 
  col = colorRamp2(myBreaks, myCol),
  top_annotation = ha.b,
  # row (gene) parameters
  cluster_rows = TRUE,
  show_row_dend = TRUE,
  #row_title = 'Statistically significant genes',
  row_title_side = 'left',
  row_title_gp = gpar(fontsize = 12,  fontface = 'bold'),
  row_title_rot = 90,
  row_names_gp = gpar(fontsize = 15, fontface = 'bold'),
  row_names_side = 'left',
  row_dend_width = unit(25,'mm'),
  show_row_names = FALSE,
  # cluster methods for rows and columns
  clustering_distance_columns = function(x) as.dist(1 - cor(t(x))),
  clustering_method_columns = 'ward.D2',
  clustering_distance_rows = function(x) as.dist(1 - cor(t(x))),
  clustering_method_rows = 'ward.D2',
  show_column_names = FALSE,
)

pdf(file ="~/My Drive/Bench/IRE1 Polysome Seq 2022/Poly_heatmap.pdf", width = 9, height = 12)
draw(heatmap.b+genelabels)

```

```{r volcano}

# ALL POLYSOME VOLCANO
#Theres a handful of Ensembl ids that are gene duplicates, keep the ones with the highest counts
o <- order(res.dds.poly.all.LFC$baseMean, decreasing=TRUE)
res.dds.poly.all.LFC <- res.dds.poly.all.LFC[o,]
dups <- duplicated(res.dds.poly.all.LFC$external_gene_name)
res.dds.poly.all.LFC <- res.dds.poly.all.LFC[!dups,]


rownames(res.dds.poly.all.LFC) <- res.dds.poly.all.LFC$external_gene_name
res.dds.poly.all.LFC <- subset(res.dds.poly.all.LFC, abs(log2FoldChange) < 9 )
res.dds.poly.all.LFC <- subset(res.dds.poly.all.LFC, padj > 1.345695e-16 )

library(EnhancedVolcano)
lab_italics = paste0("italic('", rownames(res.dds.poly.all.LFC), "')") 


selectlabels_italic <- paste0(
  "italic('",  c("Gbx1","Ucp1","Gm15793","Gm4767","Gm32828","Pdzk1","Il12rb1","Tdrd6","Snora17","Snora61", "Snora70", "Snord13", "Snora28", "Mroh8", "Snord89", "Nlrp3", "Casp12","Rab17","M1ap", "Slc1a7","Wdfy4",
                 "Hkdc1", "Casp12", "Tssk1", "Uchl1o", "Rab17", "Snora57", "Gm15793", "Adprhl1", "Card14", "Sbsn", "Foxd1", "Nanos2","Tmem207"),
  "')")


pCutoff = 0.05
FCcutoff = 0.5

poly.all.volcano <- 
  EnhancedVolcano(res.dds.poly.all.LFC, 
                  lab = lab_italics, 
                  selectLab = selectlabels_italic,
                  x = 'log2FoldChange', y = 'padj',
                  xlab = bquote(~Log[2]~ 'fold change (shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  ylim = c(0,8),
                  xlim = c(-3.5,3.5),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff, 
                  pointSize = 1.0, 
                  labSize = 5.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Gene Expression in IRE1"*alpha*"KO Polysome RNA"),
                  subtitle = "All Polysome Fractions",
                  caption = paste0('log2 FC cutoff: ', 
                                   FCcutoff, '; p-value cutoff: ', 
                                   pCutoff, '\nTotal = ', 
                                   
                                   nrow(res.dds.poly.all.LFC), ' variables')
  ) + coord_flip()#,
#legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
#legendPosition = 'bottom', legendLabSize = 14, legendIconSize = 5.0)


poly.all.volcano
ggsave(plot=last_plot(), file ="~/My Drive/Bench/IRE1 Polysome Seq 2022/polysome_volcano.pdf", width = 7, height = 10)

```

```{r heavy_volcano}


o <- order(res.dds.heavy.LFC$baseMean, decreasing=TRUE)
res.dds.heavy.LFC <- res.dds.heavy.LFC[o,]
dups <- duplicated(res.dds.heavy.LFC$external_gene_name)
res.dds.heavy.LFC <- res.dds.heavy.LFC[!dups,]


rownames(res.dds.heavy.LFC) <- res.dds.heavy.LFC$external_gene_name

res.dds.heavy.LFC <- subset(res.dds.heavy.LFC, abs(log2FoldChange) < 8)
lab_italics = paste0("italic('", rownames(res.dds.heavy.LFC), "')") 

selectlabels_italic <- paste0(
  "italic('",  c("Naaladl2", "Wdfy4", "Dock8", "Itga2", "Tns2", "Rny2", "Snord94", "Rnu11", "Rps27", "Rnu12", "Snora23", "Tet2",  "Abca8a","Rpph1","Ank1","Gm22988"),
  "')")


pCutoff = 0.05
FCcutoff = 0.5

heavy.volcano <- 
  EnhancedVolcano(res.dds.heavy.LFC, 
                  lab = lab_italics, 
                  selectLab = selectlabels_italic,
                  x = 'log2FoldChange', y = 'padj',
                  xlab = bquote(~Log[2]~ 'fold change (Shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff, 
                  pointSize = 1.0, 
                  labSize = 3.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Gene Expression IRE1"*alpha*"KO Heavy Polysome RNA"),
                  subtitle = "...",
                  caption = paste0('log2 FC cutoff: ', 
                    FCcutoff, '; p-value cutoff: ', 
                    pCutoff, '\nTotal = ', 
                    
                    nrow(res.dds.heavy.LFC), ' variables')
                  ) + coord_flip()#,
  #legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
  #legendPosition = 'bottom', legendLabSize = 14, legendIconSize = 5.0)


heavy.volcano
ggsave(plot=last_plot(), file ="~/My Drive/Bench/IRE1 Polysome Seq 2022/plots/heavy_polysome_volcano.pdf", width = 7, height = 7)

```

```{r light_volcano}

o <- order(res.dds.light.LFC$baseMean, decreasing=TRUE)
res.dds.light.LFC <- res.dds.light.LFC[o,]
dups <- duplicated(res.dds.light.LFC$external_gene_name)
res.dds.light.LFC <- res.dds.light.LFC[!dups,]


rownames(res.dds.light.LFC) <- res.dds.light.LFC$external_gene_name

#res.dds.light.LFC <- subset(res.dds.light.LFC, abs(log2FoldChange) < 8)
lab_italics = paste0("italic('", rownames(res.dds.light.LFC), "')") 



pCutoff = 0.05
FCcutoff = 0.5

light.volcano <- 
  EnhancedVolcano(res.dds.light.LFC, 
                  lab = lab_italics, 
                  #selectLab = selectlabels_italic,
                  x = 'log2FoldChange', y = 'padj',
                  xlab = bquote(~Log[2]~ 'fold change (Shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff, 
                  pointSize = 1.0, 
                  labSize = 3.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Gene Expression in IRE1"*alpha*"KO Light Polysome RNA"),
                  subtitle = "...",
                  caption = paste0('log2 FC cutoff: ', 
                    FCcutoff, '; p-value cutoff: ', 
                    pCutoff, '\nTotal = ', 
                    
                    nrow(res.dds.light.LFC), ' variables')
                  ) + coord_flip()#,
  #legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
  #legendPosition = 'bottom', legendLabSize = 14, legendIconSize = 5.0)


light.volcano
ggsave(plot=last_plot(), file ="~/My Drive/Bench/IRE1 Polysome Seq 2022/plots/light_polysome_volcano.pdf", width = 7, height = 7)





```


```{r se_volcano}


o <- order(res.dds.se.LFC$baseMean, decreasing=TRUE)
res.dds.se.LFC <- res.dds.se.LFC[o,]
dups <- duplicated(res.dds.se.LFC$external_gene_name)
res.dds.se.LFC <- res.dds.se.LFC[!dups,]


rownames(res.dds.se.LFC) <- res.dds.se.LFC$external_gene_name

#res.dds.light.LFC <- subset(res.dds.light.LFC, abs(log2FoldChange) < 8)
lab_italics = paste0("italic('", rownames(res.dds.se.LFC), "')") 



pCutoff = 0.05
FCcutoff = 0.5

light.volcano <- 
  EnhancedVolcano(res.dds.se.LFC, 
                  lab = lab_italics, 
                  #selectLab = selectlabels_italic,
                  x = 'log2FoldChange', y = 'padj',
                  xlab = bquote(~Log[2]~ 'fold change (Shrunk)'), 
                  ylab = bquote(~-Log[10]~adjusted~italic(P)),
                  boxedLabels = TRUE,
                  parseLabels = TRUE,
                  pCutoff = pCutoff, 
                  FCcutoff = FCcutoff, 
                  pointSize = 1.0, 
                  labSize = 3.0,
                  drawConnectors = TRUE,
                  widthConnectors = 0.5,
                  title = expression("Differential Gene Expression in IRE1"*alpha*"KO Total RNA"),
                  subtitle = "...",
                  caption = paste0('log2 FC cutoff: ', 
                    FCcutoff, '; p-value cutoff: ', 
                    pCutoff, '\nTotal = ', 
                    
                    nrow(res.dds.se.LFC), ' variables')
                  ) + coord_flip()#,
  #legend=c('NS','Log2 FC','Adjusted p-value', 'Adjusted p-value & Log2 FC'),
  #legendPosition = 'bottom', legendLabSize = 14, legendIconSize = 5.0)


light.volcano
ggsave(plot=last_plot(), file ="~/My Drive/Bench/IRE1 Polysome Seq 2022/plots/final/total_RNA_volcano.pdf", width = 7, height = 7)



```


```{r GO}

##################### GENE ONTOLOGY  ############################

# Create a named vector of log2 fold changes with gene symbol as names
allgenes.logfc <- res.dds.poly.all.LFC$log2FoldChange
names(allgenes.logfc) <- res.dds.poly.all.LFC$external_gene_name
# Sort the log2_fc vector in decreasing order
allgenes.logfc <- sort(allgenes.logfc, decreasing = TRUE)
#remove any NA gene 
allgenes.logfc <- na.omit(allgenes.logfc)

# Do the same for significant genes 
sig.genes.logfc <- res.dds.poly.all.LFC$log2FoldChange
names(sig.genes.logfc) <- rownames(res.dds.poly.all.LFC)
# Sort the log2_fc vector in decreasing order
sig.genes.logfc <- sort(sig.genes.logfc, decreasing = TRUE)
sig.genes.logfc <- sig.genes.logfc[names(sig.genes.logfc) %in% sig.poly$external_gene_name]
#remove any NA gene 
sig.genes.logfc <- na.omit(sig.genes.logfc)

library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(AnnotationDbi)


Poly.gse.results.all <- gseGO(geneList = allgenes.logfc,
                        OrgDb = org.Mm.eg.db,
                        keyType = "SYMBOL",
                        ont = "ALL",
                        minGSSize = 20,
                        maxGSSize = 1000,
                        pvalueCutoff = 0.05,
                        pAdjustMethod = "BH",
                        verbose = TRUE)



dotplot(Poly.gse.results.all, showCategory = 10) + labs(x = "enrichment distribution")
cnetplot(Poly.gse.results.all,  showCategory = 4,
         color.params = list(foldChange =allgenes.logfc),
         cex.params = list(category_node = 1, gene_node = 0.5),
         repel = TRUE)


Poly.gse.results.all <- pairwise_termsim(Poly.gse.results.all)
emapplot(Poly.gse.results.all, 
         showCategory = 8,
         cex.params = list(category_node = 0.8,
                           line = 0.5),
         force = 2)
GSEA.results <- data.frame(Poly.gse.results.all)
write.table(GSEA.results, file= "GSEA_results.csv", sep = ",", row.names = FALSE, quote = FALSE)

# Perform GSEA analysis (Biological Pathway "BP")
Poly.gse.results.BP <- gseGO(geneList = allgenes.logfc,
                        OrgDb = org.Mm.eg.db,
                        keyType = "SYMBOL",
                        ont = "BP",
                        minGSSize = 20,
                        maxGSSize = 1000,
                        pvalueCutoff = 0.05,
                        pAdjustMethod = "BH",
                        verbose = TRUE)

gseaplot2(Poly.gse.results.BP,  
          title = Poly.gse.results.BP$Description[2], 
          geneSetID = 2,
          base_size = 18)

gseaplot2(Poly.gse.results.all,  
          geneSetID = c(2:5, 22:23),
          pvalue_table = TRUE,
          base_size = 18,
          subplots = 1:2,
          rel_heights = c(1,0.25,1),
          color = c("#1A5354","#5A9599","#FF95A8","#F79D1E","#3F4041","#8A4198"))


ridgeplot(Poly.gse.results.BP) + labs(x = "enrichment distribution")


# Perform GSEA analysis (Molecular Function "MF")
Poly.gse.results.MF <- gseGO(geneList = allgenes.logfc,
                        OrgDb = org.Mm.eg.db,
                        keyType = "SYMBOL",
                        ont = "MF",
                        minGSSize = 20,
                        maxGSSize = 1000,
                        pvalueCutoff = 0.05,
                        pAdjustMethod = "BH",
                        verbose = TRUE)
ridgeplot(Poly.gse.results.MF) + labs(x = "enrichment distribution")
gseaplot2(Poly.gse.results.MF, title = Poly.gse.results.MF$Description[1], geneSetID = 1)

cnetplot(gse, categorySize="pvalue", foldChange=gene_list, showCategory = 3)


### TMOD
# First great a mouse gene module set from MSigDB
library(msigdbr)
msig <- msigdbr(species = "Mus musculus", category = "C2")
library(tmod)
msig <- makeTmodFromDataFrame(df=msig, 
  feature_col="gene_symbol",
  module_col="gs_id", title_col="gs_name", 
  extra_module_cols=c("gs_cat", "gs_subcat", "gs_url", 
                      "gs_exact_source", "gs_description"))

tt <- res.dds.poly.all.LFC
tt <- tt[ order(tt$padj), ]
l  <- tt$external_gene_name
#l  <- toupper(l)
tmod.res <- tmodCERNOtest(l, mset=msig)

sig.poly.tmod <- subset(res.dds.poly.all.LFC, abs(log2FoldChange) > 1)
sig.poly.tmod$colour <- ifelse(sig.poly.tmod$log2FoldChange > 0,"blue","red")
tmod.colour <- sig.poly.tmod$colour
names(tmod.colour) <- sig.poly.tmod$external_gene_name

evidencePlot(l, "M8229", mset = msig, gene.labels = sig.poly.tmod $external_gene_name)
evidencePlot(l, "M27686", mset = msig, gene.labels = sig.poly$external_gene_name)


```